fil_fungible_token/src/token/mod.rs

```diff
@@ -272,71 +318,92 @@ where
             )));
         }
 
-        let _old_state = self.state.clone();
+        let old_state = self.state.clone();
 
         self.transaction(|state, bs| {
             if spender != owner {
                 // attempt to use allowance and return early if not enough
-                state.attempt_use_allowance(&bs, spender, owner, &value)?;
+                state.attempt_use_allowance(&bs, spender, owner, value)?;
             }
-            state.change_balance_by(&bs, receiver, &value)?;
+            state.change_balance_by(&bs, receiver, value)?;
             state.change_balance_by(&bs, owner, &value.neg())?;
             Ok(())
         })?;
 
-        // TODO: call hook
-        {
-            // flush state as re-entrant call needs to see new balances
-            // self.flush()?;
+        // flush state as re-entrant call needs to see new balances
+        self.flush()?;
 
-            // call hook here
-
-            // if hook aborted, return to previous state
-            // self.state = _old_state;
-            // self.flush()?;
+        // call receiver hook
+        match self.mc.call_receiver_hook(owner, receiver, value, data) {
+            Ok(true) => {
+                // hook returned true, so we can continue
+                Ok(())
+            }
+            Ok(false) => {
+                // receiver hook aborted, revert state
+                self.state = old_state;
+                self.flush()?;
+                Err(TokenError::ReceiverHook {
+                    from: owner,
+                    to: receiver,
+                    by: spender,
+                    value: value.clone(),
+                })
+            }
+            Err(e) => {
+                // error calling receiver hook, revert state
+                self.state = old_state;
+                self.flush()?;
+                Err(e.into())
+            }
```

Comment:
Is it possible to factor out this repeated code?

Reply:
I'll defer this refactor after some more benchmarking to settle on the preferred rollback strategy